libs/@local/codec/src/harpc/wire.rs

```diff
@@ -0,0 +1,76 @@
+use core::marker::PhantomData;
+use std::io;
+
+use error_stack::{Report, ResultExt};
+use harpc_wire_protocol::{
+    codec::{Buffer, Decode, Encode},
+    request::Request,
+    response::Response,
+};
+use tokio_util::{
+    bytes::BytesMut,
+    codec::{Decoder, Encoder},
+};
+
+pub struct ProtocolEncoder<T>(PhantomData<fn() -> *const T>);
+
+impl<T> Encoder<T> for ProtocolEncoder<T>
+where
+    T: Encode,
+{
+    // we need to use io::Error here, because `Encoder` requires `From<io::Error> for Error`
+    type Error = Report<io::Error>;
+
+    fn encode(&mut self, item: T, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        let mut buffer = Buffer::new(dst);
+
+        item.encode(&mut buffer)
+            .change_context(io::Error::from(io::ErrorKind::InvalidData))
+    }
+}
+
+pub struct ProtocolDecoder<T>(PhantomData<fn() -> *const T>);
```

Comment:
The type of `ProtocolEncoder` and `ProtocolDecoder` is the same. Can't we reuse it?
```suggestion
pub struct ProtocolCodec<T>(PhantomData<fn() -> *const T>);
```

Reply:
addressed in [05d2e73](https://github.com/hashintel/hash/pull/4389/commits/05d2e73b18cbc0207217af53062812175c631af9)